core/src/main/scala/scalaz/Either.scala

```diff
@@ -263,52 +263,52 @@ sealed abstract class \/[+A, +B] extends Product with Serializable {
     }
 
   /** Ensures that the right value of this disjunction satisfies the given predicate, or returns left with the given value. */
-  def ensure[AA >: A](onLeft: => AA)(f: B => Boolean): (AA \/ B) = this match {
-    case \/-(b) => if (f(b)) this else -\/(onLeft)
-    case -\/(_) => this
-  }
+  def ensure[AA](onLeft: => AA)(f: B => Boolean)(implicit evA: A <~< AA): (AA \/ B) = Liskov.co2(evA)(this) match {
+      case either @ \/-(b) => if (f(b)) either else -\/(onLeft)
+      case either @ -\/(_) => either
+    }
 
   /** Run the given function on the left and return right with the result. */
-  def recover[BB >: B](pf: PartialFunction[A, BB]): (A \/ BB) = this match {
+  def recover[BB](pf: PartialFunction[A, BB])(implicit evB: B <~< BB): (A \/ BB) = this match {
     case -\/(a) if (pf isDefinedAt a) => \/-(pf(a))
-    case _ => this
+    case _ => Liskov.co2_2(evB)(this)
   }
 
   /** Run the given function on the left and return the result. */
-  def recoverWith[AA >: A, BB >: B](pf: PartialFunction[AA, AA \/ BB]): (AA \/ BB) = this match {
+  def recoverWith[AA, BB](pf: PartialFunction[AA, AA \/ BB])(implicit evA: A <~< AA, evB: B <~< BB): (AA \/ BB) = Liskov.lift2(evA, evB)(this) match {
     case -\/(a) if (pf isDefinedAt a) => pf(a)
-    case _ => this
+    case other => other
   }
 
   /** Compare two disjunction values for equality. */
-  def ===[AA >: A, BB >: B](x: AA \/ BB)(implicit EA: Equal[AA], EB: Equal[BB]): Boolean =
+  def ===[AA, BB](x: AA \/ BB)(implicit evA: A <~< AA, evB: B <~< BB, EA: Equal[AA], EB: Equal[BB]): Boolean =
     this match {
       case -\/(a1) => x match {
-        case -\/(a2) => Equal[AA].equal(a1, a2)
+        case -\/(a2) => Equal[AA].equal(evA(a1), a2)
         case \/-(_) => false
       }
       case \/-(b1) => x match {
-        case \/-(b2) => Equal[BB].equal(b1, b2)
+        case \/-(b2) => Equal[BB].equal(evB(b1), b2)
         case -\/(_) => false
       }
     }
 
   /** Compare two disjunction values for ordering. */
-  def compare[AA >: A, BB >: B](x: AA \/ BB)(implicit EA: Order[AA], EB: Order[BB]): Ordering =
+  def compare[AA, BB](x: AA \/ BB)(implicit evA: A <~< AA, evB: B <~< BB, EA: Order[AA], EB: Order[BB]): Ordering =
     this match {
       case -\/(a1) => x match {
-        case -\/(a2) => Order[AA].apply(a1, a2)
+        case -\/(a2) => Order[AA].apply(evA(a1), a2)
         case \/-(_) => Ordering.LT
       }
       case \/-(b1) => x match {
-        case \/-(b2) => Order[BB].apply(b1, b2)
+        case \/-(b2) => Order[BB].apply(evB(b1), b2)
         case -\/(_) => Ordering.GT
       }
     }
 
   /** Show for a disjunction value. */
-  def show[AA >: A, BB >: B](implicit SA: Show[AA], SB: Show[BB]): Cord =
-    this match {
+  def show[AA, BB](implicit evA: A <~< AA, evB: B <~< BB, SA: Show[AA], SB: Show[BB]): Cord =
```

Comment:
Discussion in #1487 says we should get rid of the subtype relationship here and just use `Show[A]` and `Show[B]`.  However, the scala compiler whines about variance if we do it now.  This will have to wait until Issue #1469 is resolved... Unless there's some scalac variance magic I don't know about.

Reply:
What exactly is the problem with variance here?

Reply:
Covariant type in invariant position.

```
[error] /Users/jacobbarber/projects/scalaz/core/src/main/scala/scalaz/Either.scala:310:21: covariant type A occurs in invariant position in type scalaz.Show[A] of value SA
[error]   def show(implicit SA: Lambda[a => Show[a]][A], SB: Show[B]): Cord =
[error]                     ^
```